File: src/services/parsers/dateParser.ts
import { addDays, addWeeks, addMonths, addYears, nextDay, setHours, setMinutes } from 'date-fns';
import { consume } from './normalize';

export interface ParsedDate {
    date: Date;
    hasExplicitTime: boolean;
}

const WEEKDAYS: { [key: string]: number } = {
    'so': 0, 'sonntag': 0, 'sunday': 0, 'sun': 0,
    'mo': 1, 'montag': 1, 'monday': 1, 'mon': 1,
    'di': 2, 'dienstag': 2, 'tuesday': 2, 'tue': 2,
    'mi': 3, 'mittwoch': 3, 'wednesday': 3, 'wed': 3,
    'do': 4, 'donnerstag': 4, 'thursday': 4, 'thu': 4,
    'fr': 5, 'freitag': 5, 'friday': 5, 'fri': 5,
    'sa': 6, 'samstag': 6, 'saturday': 6, 'sat': 6
};

/**
 * Parse relative dates (heute, morgen, übermorgen, etc)
 */
export const parseRelativeDate = (text: string): [Date | null, string] => {
    const now = new Date();
    let remaining = text;

    if (/\b(heute|today)\b/i.test(remaining)) {
        return [now, consume(remaining, /\b(heute|today)\b/i)];
    }

    if (/\b(morgen|tomorrow)\b/i.test(remaining)) {
        return [addDays(now, 1), consume(remaining, /\b(morgen|tomorrow)\b/i)];
    }

    if (/\b(übermorgen)\b/i.test(remaining)) {
        return [addDays(now, 2), consume(remaining, /\b(übermorgen)\b/i)];
    }

    if (/\b(nächste woche|next week)\b/i.test(remaining)) {
        return [addWeeks(now, 1), consume(remaining, /\b(nächste woche|next week)\b/i)];
    }

    if (/\b(in|in)\s+(\d+)\s+(tagen?|wochen?|monaten?|days?|weeks?|months?)\b/i.test(remaining)) {
        const match = remaining.match(/\b(in|in)\s+(\d+)\s+(tagen?|wochen?|monaten?|tagen?|tagen?|days?|weeks?|months?)\b/i);
        if (match) {
            const amount = parseInt(match[2], 10);
            const unit = match[3].toLowerCase();
            let date: Date = now;

            if (unit.startsWith('tag') || unit.startsWith('day')) {
                date = addDays(now, amount);
            } else if (unit.startsWith('woch') || unit.startsWith('week')) {
                date = addWeeks(now, amount);
            } else if (unit.startsWith('monat') || unit.startsWith('month')) {
                date = addMonths(now, amount);
            }

            return [date, consume(remaining, match[0])];
        }
    }

    return [null, remaining];
};

/**
 * Parse specific weekday (Montag, Monday, etc)
 */
export const parseWeekdayDate = (text: string): [Date | null, string] => {
    const now = new Date();
    let remaining = text;

    for (const [key, dayIdx] of Object.entries(WEEKDAYS)) {
        const regex = new RegExp(`\\b(am|on|next|nächsten?|nächster)?\\s*${key}\\b`, 'i');
        if (regex.test(remaining)) {
            const nextDate = nextDay(now, dayIdx as any);
            return [nextDate, consume(remaining, regex)];
        }
    }

    return [null, remaining];
};

/**
 * Parse time from text (HH:MM, 18 Uhr, 6 PM, etc)
 */
export const parseTime = (text: string): [number | null, number | null, string] => {
    let remaining = text;
    let hours: number | null = null;
    let minutes: number | null = null;

    // HH:MM or HH.MM format
    const timeMatch = remaining.match(/\b(\d{1,2})[:.](\d{2})\b/);
    if (timeMatch) {
        hours = parseInt(timeMatch[1], 10);
        minutes = parseInt(timeMatch[2], 10);
        remaining = consume(remaining, timeMatch[0]);
        return [hours, minutes, remaining];
    }

    // H am/pm format
    const ampmMatch = remaining.match(/\b(\d{1,2})\s*(uhr|am|pm|a\.m\.|p\.m\.)\b/i);
    if (ampmMatch) {
        hours = parseInt(ampmMatch[1], 10);
        const period = ampmMatch[2].toLowerCase();
        if (period.includes('pm') && hours < 12) hours += 12;
        if (period.includes('am') && hours === 12) hours = 0;
        minutes = 0;
        remaining = consume(remaining, ampmMatch[0]);
        return [hours, minutes, remaining];
    }

    // Generic time keywords
    const keywords: { regex: RegExp; h: number }[] = [
        { regex: /\b(morgens?|morning|früh|early|vormittags?|frühstück|breakfast)\b/i, h: 8 },
        { regex: /\b(mittags?|noon|lunch|mittagessen)\b/i, h: 12 },
        { regex: /\b(nachmittags?|afternoon)\b/i, h: 15 },
        { regex: /\b(abends?|evening|tonight|dinner|abendessen)\b/i, h: 19 },
        { regex: /\b(nachts?|night|midnight)\b/i, h: 23 }
    ];

    for (const kw of keywords) {
        if (kw.regex.test(remaining)) {
            hours = kw.h;
            minutes = 0;
            remaining = consume(remaining, kw.regex);
            return [hours, minutes, remaining];
        }
    }

    return [null, null, remaining];
};

/**
 * Apply time to a date
 */
export const applyTime = (date: Date, hours: number, minutes: number): Date => {
    return setHours(setMinutes(date, minutes), hours);
};

File: LumaTask/Services/NLP/DateParser.swift
import Foundation

struct DateParser {
    private let normalizer = TextNormalizer()
    private let calendar = Calendar.current

    private let weekdays: [String: Int] = [
        "so": 1, "sonntag": 1, "sunday": 1, "sun": 1,
        "mo": 2, "montag": 2, "monday": 2, "mon": 2,
        "di": 3, "dienstag": 3, "tuesday": 3, "tue": 3,
        "mi": 4, "mittwoch": 4, "wednesday": 4, "wed": 4,
        "do": 5, "donnerstag": 5, "thursday": 5, "thu": 5,
        "fr": 6, "freitag": 6, "friday": 6, "fri": 6,
        "sa": 7, "samstag": 7, "saturday": 7, "sat": 7
    ]

    func parseRelativeDate(_ text: String) -> (Date?, String) {
        let now = Date()
        let remaining = text

        if remaining.range(of: "\\b(heute|today)\\b", options: [.regularExpression, .caseInsensitive]) != nil {
            return (now, normalizer.consume(remaining, pattern: "\\b(heute|today)\\b"))
        }

        if remaining.range(of: "\\b(morgen|tomorrow)\\b", options: [.regularExpression, .caseInsensitive]) != nil {
            return (now.addingDays(1), normalizer.consume(remaining, pattern: "\\b(morgen|tomorrow)\\b"))
        }

        if remaining.range(of: "\\b(übermorgen|uebermorgen)\\b", options: [.regularExpression, .caseInsensitive]) != nil {
            return (now.addingDays(2), normalizer.consume(remaining, pattern: "\\b(übermorgen|uebermorgen)\\b"))
        }

        if remaining.range(of: "\\b(nächste woche|naechste woche|next week)\\b", options: [.regularExpression, .caseInsensitive]) != nil {
            return (now.addingWeeks(1), normalizer.consume(remaining, pattern: "\\b(nächste woche|naechste woche|next week)\\b"))
        }

        // "in X tagen/wochen/monaten"
        let relativePattern = "\\b(in)\\s+(\\d+)\\s+(tagen?|wochen?|monaten?|days?|weeks?|months?)\\b"
        if let match = remaining.firstMatch(for: relativePattern) {
            let amountStr = match.group(2) ?? "0"
            let amount = Int(amountStr) ?? 0
            let unit = (match.group(3) ?? "").lowercased()

            var date = now
            if unit.hasPrefix("tag") || unit.hasPrefix("day") {
                date = now.addingDays(amount)
            } else if unit.hasPrefix("woch") || unit.hasPrefix("week") {
                date = now.addingWeeks(amount)
            } else if unit.hasPrefix("monat") || unit.hasPrefix("month") {
                date = now.addingMonths(amount)
            }

            return (date, normalizer.consume(remaining, pattern: relativePattern))
        }

        return (nil, remaining)
    }

    func parseWeekdayDate(_ text: String) -> (Date?, String) {
        let now = Date()
        let remaining = text

        for (key, weekdayIndex) in weekdays {
            let pattern = "\\b(am|on|next|nächsten?|nächster|naechsten?)?\\s*\(NSRegularExpression.escapedPattern(for: key))\\b"
            if remaining.range(of: pattern, options: [.regularExpression, .caseInsensitive]) != nil {
                if let nextDate = calendar.nextDate(after: now, matching: DateComponents(weekday: weekdayIndex), matchingPolicy: .nextTime) {
                    return (nextDate, normalizer.consume(remaining, pattern: pattern))
                }
            }
        }

        return (nil, remaining)
    }
}
